src/Microsoft.AspNetCore.JsonPatch/Internal/DynamicObjectAdapter.cs

```diff
@@ -94,6 +94,11 @@ public class DynamicObjectAdapter : IAdapter
                 return false;
             }
 
+            if (!TryRemove(target, segment, contractResolver, out errorMessage))
+            {
+                return false;
+            }
+
```

Comment:
The `TrySetDynamicObjectProperty` method, either creates a new entry in the underlying dictionary in the `DynamicObject` class or sets the existing property to the new value (in effect replacing the existing property). It shouldn't be necessary to remove it.

Reply:
Right, but the issue as I described in #110 is that simply replacing the value using `TrySetDynamicObjectProperty` means that "replace" operations are not distinguishable from "add" operations. From RFC 6902, the operation should act as a remove operation (using `TrySetDynamicObjectProperty` to set the property to `null`) followed immediately by an add operation.

This is how you had [it implemented originally](https://github.com/aspnet/JsonPatch/blob/patch/1.0.5/src/Microsoft.AspNetCore.JsonPatch/Adapters/ObjectAdapter.cs#L725) (setting `null` before setting the replacement value), so I believe this is an unintended breaking change. The issue I'm having now concerns an operation that I'm fairly sure I've tried before and didn't give me problems when I was using JSON Patch v1.0.0.

Reply:
"replace" is different from "add" because when replacing, the new value needs to be the same type as the original value assigned to the key.

Reply:
Sure, but my PR would not affect those semantics. My changes don't affect the final value that gets set for the property, it just temporarily replaces it with null beforehand (signalling property removal).